Continuous backlog population

pwojcikdev opened the pull request:
Backlog population is a process in which a node scans all accounts in the ledger, with or without any confirmed blocks, and forwards (activates) those accounts which do not have all their blocks confirmed to election scheduler for prioritization and eventual queuing in proper bucket. It is necessary to do this periodically, because the amount of space in each bucket is limited (currently ~2000 entries) and number of accounts needing confirmations can be much higher than that, especially during bootstrap or network spam attack.

The problem with current implementation is that this process runs every 5 minutes and scans the whole ledger at once, leading to situations where we run out of accounts to prioritize before the next run has started. This is especially visible during bootstrapping, a graph showing such situation is included below. We can clearly see the bumps in AEC occupancy where prioritization queue is filled, followed by periods of idleness when priority queue is emptied:

<img width="2955" alt="Screen Shot 2022-11-15 at 21 15 39" src="https://user-images.githubusercontent.com/3044353/202019992-12882891-b080-4bd9-8771-ea2617734758.png"> 

This PR fixes that by modifying the way the ledger scan is done. Instead of 5 minute interval, we run the scan all the time (unless disabled by setting `frontiers_confirmation = disabled` node config setting), but we throttle the rate at which the scan is done to limit consumption of node resources. The rate and frequency is controlled by two new node-config.toml settings: `backlog_scan_batch_size` and `backlog_scan_frequency`. By default it scans 10000 accounts per second divided into 10 batches, so 1000 accounts per batch. This is rather conservative and should be later adjusted with feedback from beta node operators (before this PR we dit it in batches of 64k).

The result of this PR is the AEC that stays full almost all the time (except the initial phase of the bootstrap):

<img width="2952" alt="Screen Shot 2022-11-15 at 21 15 53" src="https://user-images.githubusercontent.com/3044353/202032045-7074f99d-210a-4962-97ac-bf1f84d21790.png">

pwojcikdev committed:
Continuous backlog population

pwojcikdev committed:
Add backlog population stats

pwojcikdev committed:
Invert dependencies

pwojcikdev committed:
Add backlog population test

pwojcikdev committed:
Rework backlog config

pwojcikdev committed:
Add node config options

pwojcikdev committed:
Fix compilation

dsiganos commented:
It looks like the unit test 'request_aggregator.cannot_vote' failed 4 times.

pwojcikdev commented:
@dsiganos I see, there seems to be two tests that are still failing, I'm looking into that. Appears to only break on GH runners, so a bit annoying to debug.

dsiganos commented:
It fails when the system is under heavy load. Starting a parallel build from scratch seems to make it crash.
I am currently fixing the qwahzi unit test and then I will come back to this.

I got this crash on my laptop:
```
[ RUN      ] request_aggregator.cannot_vote
/Users/ds/CLionProjects/nano-node/nano/core_test/request_aggregator.cpp:501: Failure
Value of: system.poll ()
  Actual: Deadline expired
Expected:
zsh: segmentation fault  ./core_test --gtest_repeat=1000 --gtest_filter=request_aggregator.cannot_vote
```

dsiganos commented on nano/node/backlog_population.hpp:
```diff
@@ -16,43 +20,58 @@ class backlog_population final
+		/** Number of accounts per second to process. Number of accounts per single batch is this value divided by `frequency` */
+		unsigned rate;
+
+		/** Number of batches to run per second. Batches run in 1 second / `frequency` intervals */
```
It would be nice to document the unit used. Hz, MHz, etc. It seems to be Hz.
And would it be more accurate to call it sleep frequency?

If this is counting Hz then the sleep period cannot be less than 1 second. A minor point but making it Mhz would give us greater flexibility.

dsiganos commented on nano/node/backlog_population.cpp:
```diff
@@ -49,48 +50,81 @@ void nano::backlog_population::notify ()
-			next = account.number () + 1;
+			auto transaction = store.tx_begin_read ();
+
+			auto count = 0;
```
Very minor comment. The count being of type auto doesn't add anything in my opinion. It just generates questions, like what size will it be? I think it would be more easily readable to just set it to the type intended.

dsiganos commented on nano/core_test/backlog.cpp:
```diff
@@ -0,0 +1,105 @@
+ * Creates `count` 1 raw sends from genesis to unique accounts and corresponding open blocks.
+ * The genesis chain is then confirmed, but leaves open blocks unconfirmed.
+ */
+std::vector<std::shared_ptr<nano::block>> setup_independent_blocks (nano::test::system & system, nano::node & node, int count)
```
This could return block_list_t

dsiganos commented on nano/core_test/backlog.cpp:
```diff
@@ -0,0 +1,105 @@
+/*
+ * Creates `count` 1 raw sends from genesis to unique accounts and corresponding open blocks.
+ * The genesis chain is then confirmed, but leaves open blocks unconfirmed.
+ */
```
The comment could say:
"The list of unconfirmed open blocks is returned"

dsiganos commented on nano/core_test/backlog.cpp:
```diff
@@ -0,0 +1,105 @@
+					.build_shared ();
+		latest = send->hash ();
+
+		auto open = builder
```
We want to discourage use of legacy blocks in new code. Ideally this would be a state block.

dsiganos commented on nano/core_test/backlog.cpp:
```diff
@@ -0,0 +1,105 @@
+	auto blocks = setup_independent_blocks (system, node, 256);
+
+	// Checks if `activated` set contains all accounts we previously set up
+	auto sum_all_activated = [&] () {
```
Wouldn't it be simpler and more accurate to loop through all blocks and assert that:
activated.count() == 1

dsiganos commented on nano/node/nodeconfig.cpp:
```diff
@@ -128,6 +128,8 @@ nano::error nano::node_config::serialize_toml (nano::tomlconfig & toml) const
 	toml.put ("frontiers_confirmation", serialize_frontiers_confirmation (frontiers_confirmation), "Mode controlling frontier confirmation rate.\ntype:string,{auto,always,disabled}");
 	toml.put ("max_queued_requests", max_queued_requests, "Limit for number of queued confirmation requests for one channel, after which new requests are dropped until the queue drops below this value.\ntype:uint32");
 	toml.put ("rep_crawler_weight_minimum", rep_crawler_weight_minimum.to_string_dec (), "Rep crawler minimum weight, if this is less than minimum principal weight then this is taken as the minimum weight a rep must have to be tracked. If you want to track all reps set this to 0. If you do not want this to influence anything then set it to max value. This is only useful for debugging or for people w ...
+	toml.put ("backlog_scan_rate", backlog_scan_rate, "Number of accounts per second to process when doing backlog population scan. Increasing this value will help unconfirmed frontiers get into election prioritization queue faster, however it will also increase resource usage. \ntype:unit");
```
Typo: unit -> uint

dsiganos commented:
I left a number of minor comments but it looks good to me overall.

clemahieu committed:
Fix unit test active_transactions.confirm_election_by_request which expects that it will not receive a vote for send1 because it has not made such a request, however, if the election is still or recently active, it may receive a broadcast vote before it makes a request.
Check that the election has ended on node1 and allow some time for in-flight votes broadcasts to finish before starting node2.

dsiganos commented on nano/core_test/active_transactions.cpp:
```diff
@@ -46,6 +46,10 @@ TEST (active_transactions, confirm_election_by_request)
 
+	// Wait for the election to be removed and give time for any in-flight vote broadcasts to settle
+	ASSERT_TIMELY (5s, node1.active.empty ());
+	WAIT (250ms);
```
This seems like a hack. On a very slow computer, 250 ms might not be enough for the network to settle.
Ideally, we want something more concrete.

pwojcikdev replied:
I would use a higher wait period, personally I always use 1 second. While tracking those generated votes through vote processor, network stack etc could be correct, it's not practical or worth spending time on. A longer wait time here should be fine.

clemahieu committed:
Adding review edits.

clemahieu committed:
Renaming config option per review comments.
